File: src/bridge.cc
#include <cstdio>

#include "yydb/include/bridge.h"
#include "yydb/src/bridge.rs.h"

namespace yydb {
    /* Logging */

    void mysql_log_write(std::int32_t level, rust::Str msg) {
        mysql_log_write_raw(level, msg.data(), msg.size());
    }

    /* End of logging */

    /* Lifecycle */

    int ha_yydb_core_init() {
        rust_init();

        // do other stuff
        return 0;
    }

    int ha_yydb_core_deinit() {
        rust_deinit();

        // do other stuff
        return 0;
    }

    /* End of lifecycle */

    /* Table */
    std::uint64_t ha_yydb_open_table(const char* name) {
        rust::Str name_str(name);
        return open_table(name_str);
    }

    void ha_yydb_close_table(std::uint64_t table_id) {
        close_table(table_id);
    }

    void ha_yydb_insert_row(uint64_t table_id, uint64_t key, const u_char* row, uint length) {
        insert_row(table_id, key, (const uint8_t*)row, length);
    }

    void ha_yydb_update_row(uint64_t table_id, uint64_t key, const u_char* old, const u_char* row, uint length) {
        update_row(table_id, key, (const uint8_t*)old, (const uint8_t*)row, length);
    }
    /* End of table */
}

File: src/bridge.rs
use crate::core::bridge::*;
use crate::*;

#[cxx::bridge(namespace = "yydb")]
pub mod ffi {
    // Rust types and signatures exposed to C++.
    extern "Rust" {
        // Init the rust part of the library.
        pub fn rust_init();

        // Deinit the rust part of the library.
        pub fn rust_deinit();

        // Open a table by name.
        pub fn open_table(table_name: &str) -> u64;

        // close a table by id.
        pub fn close_table(id: u64);

        // insert a row to a table.
        pub unsafe fn insert_row(table_id: u64, key: u64, data: *const u8, len: u32);

        // update a row to a table.
        pub unsafe fn update_row(
            table_id: u64,
            key: u64,
            data: *const u8,
            new_data: *const u8,
            len: u32,
        );

        // delete a row to a table.
        pub unsafe fn delete_row(table_id: u64, key: u64);

        // init round
        pub fn rnd_init(table_id: u64);

        // end round
        pub fn rnd_end(table_id: u64);

        // read next row
        pub unsafe fn rnd_next(table_id: u64, buf: *mut u8, len: u32) -> i32;

        // put hex data to log
        pub unsafe fn put_hex(data: *const u8, len: u32);

        // delete a table by name.
        pub fn delete_table(table_name: &str);
    }

    // C++ types and signatures exposed to Rust.
    unsafe extern "C++" {
        include!("yydb/include/bridge.h");

        // write log to mysql log
        pub fn mysql_log_write(level: i32, message: &str);
    }
}
